daemon/list.go

```diff
@@ -580,18 +581,45 @@ func includeContainerInList(container *container.Snapshot, filter *listContext)
 // refreshImage checks if the Image ref still points to the correct ID, and updates the ref to the actual ID when it doesn't
 func (daemon *Daemon) refreshImage(ctx context.Context, s *container.Snapshot, filter *listContext) (*types.Container, error) {
 	c := s.Container
-	tmpImage := s.Image // keep the original ref if still valid (hasn't changed)
-	if tmpImage != s.ImageID {
-		img, err := daemon.imageService.GetImage(ctx, tmpImage, imagetypes.GetImageOpts{})
-		if err != nil && !errdefs.IsNotFound(err) {
-			return nil, err
-		}
-		if err != nil || img.ImageID() != s.ImageID {
-			// ref changed, we need to use original ID
-			tmpImage = s.ImageID
+
+	// s.Image is the image reference passed by the user to create an image
+	//         can be a:
+	//         - name (like nginx, ubuntu:latest, docker.io/library/busybox:latest),
+	//         - truncated ID (abcdef),
+	//         - full digest (sha256:abcdef...)
+	//
+	// s.ImageID is the ID of the image that s.Image resolved to at the time
+	// of the container creation. It's always a full digest.
+
+	// If these match, there's nothing to refresh.
+	if s.Image == s.ImageID {
+		return &c, nil
+	}
+
+	// Check if the image reference still resolves to the same digest.
+	img, err := daemon.imageService.GetImage(ctx, s.Image, imagetypes.GetImageOpts{})
+
+	// If the image is no longer found or can't be resolved for some other
+	// reason. Update the Image to the specific ID of the original image it
+	// resolved to when the container was created.
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			log.G(ctx).WithError(err).WithFields(logrus.Fields{
+				"containerID": c.ID,
+				"image":       s.Image,
+				"imageID":     s.ImageID,
+			}).Warn("failed to resolve container image")
```

Comment:
Definitely not a blocker; we could include the `WithError()` in the fields (it uses "error" as key);

```suggestion
			log.G(ctx).WithFields(logrus.Fields{
				"error":       err,
				"containerID": c.ID,
				"image":       s.Image,
				"imageID":     s.ImageID,
			}).Warn("failed to resolve container image")
```

Reply:
Ah, right. Even better, we can use `logrus.ErrorKey` instead of hardcoding the `"error"`.

Reply:
Yeah, somewhat trying to avoid using that one, as we're trying to move away from logrus imports

And it's unlikely that const will ever change

Reply:
What's wrong with `WithError`?

Reply:
Nothing specifically wrong with `WithError`, but doing `WithError().WithFields()` means we're doing `WithFields` twice in a row (and that one is known to have some overhead)

Reply:
Ah, yes, forgot we still use `logrus.Fields` currently (soon we can replace those).

Not a big deal in either case! There's gonna be a fair share of find and replace needed to get it all out.